### ISSUE TITLE
Corrosion does not support clang++(clang with gnu cli) on windows

### ISSUE THREAD

**Issue by trondhe** [corrosion-rs/corrosion#255]
Using clang++ on windows which is something I prefer as it does not require to run vcvars.bat file which is normally used for cl.exe.
I also use it for the gcc version of the cli which is more readable to me.
CMake reports on this version on the configure step with
`The CXX compiler identification is Clang 14.0.5 with GNU-like command-line`

Defaulting to clang++ results in `cmake/Corrosion.cmake:525` to fail with 
`message(FATAL_ERROR "Unknown windows environment - Can't determine implib name")`

If this is bypassed by commenting this line out and setting it with one of the other option used with msvc, it fails first on trying to compile the winapi crate at some point with one of the error which looks like:

```
error: linking with `C:/Programs/LLVM/bin/clang++.exe` failed: exit code: 1
  |
  = note: "C:/Programs/LLVM/bin/clang++.exe" "/NOLOGO" "C:\\Users\\trond\\AppData\\Local\\Temp\\rustcjoL1jk\\symbols.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.0.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.1.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.10.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.11.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.12.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.13.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.14.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.15.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.2.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.3.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.4.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.5.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.6.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.7.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.8.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.build_script_build.3f37258f-cgu.9.rcgu.o" "F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.5g1cfhhbm1svvnmj.rcgu.o" "/LIBPATH:F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\deps" "/LIBPATH:C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd-fdaa63d6970be740.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libpanic_unwind-a684236a9da37ec6.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_demangle-85c151aafb35bf5c.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd_detect-7dbe75e48c9029c0.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libhashbrown-78044ac8790346aa.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libminiz_oxide-8743285c70baade8.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libadler-b260ead09cc3ca75.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-200564555b6c364f.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libunwind-60e4ae6d068e334f.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-d74559c1731cb9bf.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liblibc-3d51e2979c44c27c.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liballoc-268ac0a196c81649.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_core-e74895db82cfb265.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcore-d805a250e2b7e197.rlib" "C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-3f507ba7adac1c2a.rlib" "kernel32.lib" "advapi32.lib" "userenv.lib" "kernel32.lib" "ws2_32.lib" "bcrypt.lib" "msvcrt.lib" "legacy_stdio_definitions.lib" "/NXCOMPAT" "/LIBPATH:C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "/OUT:F:/repos/rusty_cmake/build/default/./cargo/build\\debug\\build\\scratch-6fd777bf549032ed\\build_script_build-6fd777bf549032ed.exe" "/OPT:REF,NOICF" "/DEBUG" "/NATVIS:C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:C:\\Users\\trond\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libstd.natvis"
  = note: clang++: error: no such file or directory: '/NOLOGO'
          clang++: error: no such file or directory: '/LIBPATH:F:/repos/rusty_cmake/build/default/./cargo/build\debug\deps'
          clang++: error: no such file or directory: '/LIBPATH:C:\Users\trond\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib'
          clang++: error: no such file or directory: 'kernel32.lib'
          clang++: error: no such file or directory: 'advapi32.lib'
          clang++: error: no such file or directory: 'userenv.lib'
          clang++: error: no such file or directory: 'kernel32.lib'
          clang++: error: no such file or directory: 'ws2_32.lib'
          clang++: error: no such file or directory: 'bcrypt.lib'
          clang++: error: no such file or directory: 'msvcrt.lib'
          clang++: error: no such file or directory: 'legacy_stdio_definitions.lib'
          clang++: error: no such file or directory: '/NXCOMPAT'
          clang++: error: no such file or directory: '/LIBPATH:C:\Users\trond\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib'
          clang++: error: no such file or directory: '/OUT:F:/repos/rusty_cmake/build/default/./cargo/build\debug\build\scratch-6fd777bf549032ed\build_script_build-6fd777bf549032ed.exe'
          clang++: error: no such file or directory: '/OPT:REF,NOICF'
          clang++: error: no such file or directory: '/DEBUG'
          clang++: error: no such file or directory: '/NATVIS:C:\Users\trond\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\etc\intrinsic.natvis'
          clang++: error: no such file or directory: '/NATVIS:C:\Users\trond\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\etc\liballoc.natvis'
          clang++: error: no such file or directory: '/NATVIS:C:\Users\trond\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\etc\libcore.natvis'
          clang++: error: no such file or directory: '/NATVIS:C:\Users\trond\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\etc\libstd.natvis'
```

where the source of this is
```
FAILED: src/CMakeFiles/_cargo-build_rusty_code F:/repos/rusty_cmake/build/default/src/CMakeFiles/_cargo-build_rusty_code
cmd.exe /C "cd /D F:\repos\rusty_cmake\src\rusty_code && C:\Programs\CMake\bin\cmake.exe -E env CARGO_TARGET_X86_64_PC_WINDOWS_MSVC_LINKER=C:/Programs/LLVM/bin/clang++.exe CXX_x86_64-pc-windows-msvc=C:/Programs/LLVM/bin/clang++.exe CORROSION_BUILD_DIR=F:/repos/rusty_cmake/build/default/src CARGO_BUILD_RUSTC=C:/Users/trond/.rustup/toolchains/stable-x86_64-pc-windows-msvc/bin/rustc.exe C:/Users/trond/.rustup/toolchains/stable-x86_64-pc-windows-msvc/bin/cargo.exe rustc --lib --target=x86_64-pc-windows-msvc --package rusty_code --manifest-path F:/repos/rusty_cmake/src/rusty_code/Cargo.toml --target-dir F:/repos/rusty_cmake/build/default/./cargo/build -- -Cdefault-linker-libraries=yes"
```

Not sure the best way to solve this, adding `clang-win-gnulike` and `clang-win-msvclike` seems like the worst solution.  
Perhaps something that splits up `cli` and `platform`.
Now I try to avoid mingw myself, but isnt it very `gcc` like for what all cmake cares for? As in, does cmake treat mingw almost identically to gcc on linux, of sorts?

If so, perhaps splitting up, `platform`, `compiler` and `cli` intro three different variables and then build up the different setup types from these could be a reasonable way?

**Comment by jschwe** [comment]
> Defaulting to clang++ results in cmake/Corrosion.cmake:525 to fail with
message(FATAL_ERROR "Unknown windows environment - Can't determine implib name")

Looks like a previous refactoring of mine introduced this. https://github.com/corrosion-rs/corrosion/pull/256 should fix this error.

> where the source of this is

> ```
> FAILED: src/CMakeFiles/_cargo-build_rusty_code F:/repos/rusty_cmake/build/default/src/CMakeFiles/_cargo-build_rusty_code
> cmd.exe /C "cd /D F:\repos\rusty_cmake\src\rusty_code && C:\Programs\CMake\bin\cmake.exe -E env CARGO_TARGET_X86_64_PC_WINDOWS_MSVC_LINKER=C:/Programs/LLVM/bin/clang++.exe CXX_x86_64-pc-windows-msvc=C:/Programs/LLVM/bin/clang++.exe CORROSION_BUILD_DIR=F:/repos/rusty_cmake/build/default/src CARGO_BUILD_RUSTC=C:/Users/trond/.rustup/toolchains/stable-x86_64-pc-windows-msvc/bin/rustc.exe C:/Users/trond/.rustup/toolchains/stable-x86_64-pc-windows-msvc/bin/cargo.exe rustc --lib --target=x86_64-pc-windows-msvc --package rusty_code --manifest-path F:/repos/rusty_cmake/src/rusty_code/Cargo.toml --target-dir F:/repos/rusty_cmake/build/default/./cargo/build -- -Cdefault-linker-libraries=yes"
> ```

Corrosion supports two scenarios for linking:
1. cargo invokes the linker
2. CMake invokes the linker

It looks like in your case cargo invokes the linker and corrosion sets it to `CARGO_TARGET_X86_64_PC_WINDOWS_MSVC_LINKER=C:/Programs/LLVM/bin/clang++.exe` because that is the CMake CXX compiler. In this case cargo is responsible for invoking the linker and judging from the error message cargo expects cl.exe cli for the msvc abi. This would need to be detected/fixed upstream on the rust side. A possible workaround might be to set the linker to `clang-cl` but keep the compiler as `clang++`. I don't have much experience with windows, so I can't really say if this would be a good workaround.

**Comment by jschwe** [comment]
I suspect this is the same issue as https://github.com/rust-lang/rust/issues/106454. Rust matches on the linker name and `clang++` != `clang` so the linker detection falls back to the target default, which is probably MSVC for the platform. 
I hope to get a fix included in time for Rust 1.68.

**Comment by jschwe** [comment]
Actually, this issue should be fixed since f03517a98ccd69dd33ce115ac0d63aab5cc4398a or release v0.3.1.
I also added a job to the CI to test the combination of windows with the msvc abi + ninja + clang. If you still encounter this issue please reopen the issue.
